bindings/go/oci/repository.go

```diff
@@ -0,0 +1,690 @@
+	// The blob.ReadOnlyBlob returned will always be an OCI Layout, readable by oci.NewFromTar.
+	// For more information on the download procedure, see NewOCILayoutWriter.
+	DownloadResource(ctx context.Context, res *descriptor.Resource) (content blob.ReadOnlyBlob, err error)
+}
+
+// Resolver defines the interface for resolving references to OCI stores.
+type Resolver interface {
+	// StoreForReference resolves a reference to a Store.
+	// Each component version can resolve to a different store.
+	StoreForReference(ctx context.Context, reference string) (Store, error)
+
+	// ComponentVersionReference returns a unique reference for a component version.
+	ComponentVersionReference(component, version string) string
+}
+
+// Store defines the interface for interacting with an OCI store.
+type Store interface {
+	content.ReadOnlyStorage
+	content.Pusher
+	content.TagResolver
+	content.Tagger
+}
+
+// Repository implements the ComponentVersionRepository interface using OCI registries.
+// Each component version is stored in a separate OCI repository.
+type Repository struct {
+	scheme *runtime.Scheme
+
+	// localBlobMemory temporarily stores local blobs until they are added to a component version.
+	// thus any local blob added with AddLocalResource will be added to the memory until
+	// AddComponentVersion is called with a reference to that resource.
+	// Note that Store implementations are expected to either allow orphaned LocalBlobs or
+	// regularly issue an async garbage collection to remove them.
+	localBlobMemory LocalBlobMemory
+
+	// resolver resolves component version references to OCI stores.
+	resolver Resolver
+}
+
+// RepositoryFromResolverAndMemory creates a new Repository instance.
+// This is a convenience function that uses the new options pattern.
+func RepositoryFromResolverAndMemory(resolver Resolver, memory LocalBlobMemory) *Repository {
+	repo, err := NewRepository(
+		WithResolver(resolver),
+		WithLocalBlobMemory(memory),
+	)
+	if err != nil {
+		panic(err)
+	}
+	return repo
+}
+
+var _ ComponentVersionRepository = (*Repository)(nil)
+
+// AddLocalResource adds a local resource to the repository.
+func (repo *Repository) AddLocalResource(
+	ctx context.Context,
+	component, version string,
+	resource *descriptor.Resource,
+	content blob.ReadOnlyBlob,
+) (newRes *descriptor.Resource, err error) {
+	done := logOperation(ctx, "add local resource",
+		slog.String("component", component),
+		slog.String("version", version),
+		slog.String("resource", resource.Name))
+	defer done(err)
+
+	reference := repo.resolver.ComponentVersionReference(component, version)
+	store, err := repo.resolver.StoreForReference(ctx, reference)
+	if err != nil {
+		return nil, err
+	}
+
+	access, err := getLocalBlobAccess(repo.scheme, resource)
+	if err != nil {
+		return nil, err
+	}
+
+	contentSizeAware, ok := content.(blob.SizeAware)
+	if !ok {
+		return nil, fmt.Errorf("content does not implement blob.SizeAware interface, size cannot be inferred")
```

Comment:
Here and everywhere else, no need to use `Errorf` without a format. It's considered a wasteful operation. :)

Reply:
I wonder why the linter didnt shout at me....

Reply:
Good question. Normally it would have. Maybe something is disabled.

Reply:
Started looking at the linters and i think they are still quite lenient: https://github.com/open-component-model/open-component-model/pull/55 this should enforce better code quality